### DISCUSSION TITLE
can opa team provide some code sample about using if else loop in a rule?

### DISCUSSION THREAD

**Discussion by zhoupoko2000** [open-policy-agent/community#85] [OPA and Rego] [answered]
I try to make below code work. basically, this requires embedding a if else loop inside a rule block...but I keep getting error no matter how I tweak the code...any suggestion here?

**the error I got in rego playground is:** 

> 1 error occurred: policy.rego:45: rego_parse_error: unexpected { token: expected \n or ; or }

> 	patch := [add_new_patch] {

> 

**where input is:**

`{
    "message": "world"
}
`
**the code is:** 
`package play

default hello = false

make_new_patch(path, value) = patchCode {

	patchCode = {
    	"op": "make",
        "path": path,
        "value": value
    }

}

add_new_patch(path, value) = patchCode {

	patchCode = {
    	"op": "add",
        "path": path,
        "value": value,
        "comment": "not make"
    }

}

hello[patch] {
	patch := [add_new_patch] {
    	input.message == "add"
    } else := [make_new_patch]
}`

**Comment by anderseknert** [comment, marked as answer]
Both `add_new_patch` and `make_new_patch` are functions, but you are referencing them as if they were rules. I can't say I fully understand what you're trying to accomplish, but if you just want to make the function call conditionally, something like this would do it:

```rego
package play

make_new_patch(path, value) = patchCode {
    patchCode = {
        "op": "make",
        "path": path,
        "value": value
    }
}

add_new_patch(path, value) = patchCode {
    patchCode = {
        "op": "add",
        "path": path,
        "value": value,
        "comment": "not make"
    }
}

hello = add_new_patch("some_path", "some_value") {
	input.message == "add"
} else = make_new_patch("other_path", "other_value")

```

**Reply by zhoupoko2000** [reply]
hey Thanks.
I tried to figure out how to embed if else block within a rule or function. below seems work now.

> package play

make_new_patch(path, value) = patchCode {
    patchCode = {
        "op": "make",
        "path": path,
        "value": value
    }
}

add_new_patch(path, value) = patchCode {
    patchCode = {
        "op": "add",
        "path": path,
        "value": value,
        "comment": "not make"
    }
}

has_spec(obj) {
	obj.spec
}

hello = add_new_patch("some_path", "some_value") {
input.message == "add"
} else = make_new_patch("other_path", "other_value")

patches[hello]{
	has_spec(input.request)

}
